promunifi/clients.go

```diff
@@ -0,0 +1,126 @@
+package promunifi
+
+import (
+	"github.com/prometheus/client_golang/prometheus"
+	"golift.io/unifi"
+)
+
+type uclient struct {
+	Anomalies         *prometheus.Desc
+	BytesR            *prometheus.Desc
+	CCQ               *prometheus.Desc
+	Noise             *prometheus.Desc
+	RoamCount         *prometheus.Desc
+	RSSI              *prometheus.Desc
+	RxBytes           *prometheus.Desc
+	RxBytesR          *prometheus.Desc
+	RxPackets         *prometheus.Desc
+	RxRate            *prometheus.Desc
+	Signal            *prometheus.Desc
+	TxBytes           *prometheus.Desc
+	TxBytesR          *prometheus.Desc
+	TxPackets         *prometheus.Desc
+	TxPower           *prometheus.Desc
+	TxRate            *prometheus.Desc
+	Uptime            *prometheus.Desc
+	WifiTxAttempts    *prometheus.Desc
+	WiredRxBytes      *prometheus.Desc
+	WiredRxBytesR     *prometheus.Desc
+	WiredRxPackets    *prometheus.Desc
+	WiredTxBytes      *prometheus.Desc
+	WiredTxBytesR     *prometheus.Desc
+	WiredTxPackets    *prometheus.Desc
+	DpiStatsApp       *prometheus.Desc
+	DpiStatsCat       *prometheus.Desc
+	DpiStatsRxBytes   *prometheus.Desc
+	DpiStatsRxPackets *prometheus.Desc
+	DpiStatsTxBytes   *prometheus.Desc
+	DpiStatsTxPackets *prometheus.Desc
+}
+
+func descClient(ns string) *uclient {
+	if ns += "_client_"; ns == "_client_" {
+		ns = "client_"
+	}
+
+	labels := []string{"id", "mac", "user_id", "site_id", "site_name",
+		"network_id", "ap_mac", "gw_mac", "sw_mac", "ap_name", "gw_name",
+		"sw_name", "radio_name", "radio", "radio_proto", "name", "channel",
+		"vlan", "ip", "essid", "bssid", "radio_desc"}
+
+	return &uclient{
+		Anomalies:         prometheus.NewDesc(ns+"anomalies_total", "Client Anomalies", labels, nil),
+		BytesR:            prometheus.NewDesc(ns+"bytes_rate", "Client Data Rate", labels, nil),
+		CCQ:               prometheus.NewDesc(ns+"ccq", "Client Connection Quality", labels, nil),
+		Noise:             prometheus.NewDesc(ns+"noise", "Client AP Noise", labels, nil),
+		RoamCount:         prometheus.NewDesc(ns+"roam_total", "Client Roam Counter", labels, nil),
+		RSSI:              prometheus.NewDesc(ns+"rssi", "Client RSSI", labels, nil),
+		RxBytes:           prometheus.NewDesc(ns+"bytes_rx_total", "Client Receive Bytes", labels, nil),
+		RxBytesR:          prometheus.NewDesc(ns+"bytes_rx_rate", "Client Receive Data Rate", labels, nil),
+		RxPackets:         prometheus.NewDesc(ns+"packets_rx_total", "Client Receive Packets", labels, nil),
+		RxRate:            prometheus.NewDesc(ns+"rate", "Client Receive Rate", labels, nil),
```

Comment:
Probably Mbps. This is the wifi transmission rate. ie. 450Mbps or 870Mbps. Honestly I haven't sleuthed the data that much. Probably need to do more of that, but I usually iron out the kinks once I start making graphs. I guess I should start doing that.... (because then I'll need to know units)

Reply:
So, probably 
```suggestion
		RxRate:            prometheus.NewDesc(ns+"rate_mbps", "Client Receive Rate", labels, nil),
```

Of course, this is why we have all these conventions in Prometheus. Less guesswork. :smile_cat:

Reply:
I'm actually not sure what unit this is. Maybe Mbps?
```
      "rx_rate": 135000,
      "rx_rate": 13000,
      "rx_rate": 351000,
      "rx_rate": 585000,
      "rx_rate": 650000,
      "rx_rate": 65000,
      "rx_rate": 72109,
      "rx_rate": 585000,
      "rx_rate": 351000,
      "rx_rate": 121570,
      "rx_rate": 526500,
      "rx_rate": 121500,
      "rx_rate": 104000,
      "rx_rate": 585000,
      "rx_rate": 1000,
      "rx_rate": 150000,
      "rx_rate": 144000,
      "rx_rate": 39000,
      "rx_rate": 135000,
```

Reply:
Looks like bps to me. Maybe `bits_per_second` / `bps`.

Reply:
If it was kbps, that would make the slowest connection (roughly) 1mbps and the fastest connection 650mbps.. that seems to fit what my WAPs (and clients) can do.. Does kbps seem right to you?

Reply:
Also I've updated many of the metric names if you don't mind giving them another once-over.

Reply:
Ahh, you're right, it is kbps, 1000 would be a 1mbps basic rate.

Reply:
It is probably worth normalizing the data, so multiply by 1000 to make the output bps. There's no need to worry about accuracy, as the data is all float64, and it just moves the exponent around.

Reply:
I'm going to do the unthinkable and divide by zero to make it mbps. I feel like port speeds are normally expressed in mbps and this just feels more accurate. It makes it match the wired ports speeds too, 10/100/1000/etc.

Reply:
Like I said, it doesn't affect real accuracy. But it's more about keeping the [base units](https://prometheus.io/docs/practices/naming/#base-units) the same. Normally we try and make everything bytes, but networking likes to use bits as the base unit for historical reasons.